framework/src/main/java/org/checkerframework/framework/util/QualifierKindHierarchy.java

```diff
@@ -0,0 +1,874 @@
+     */
+    protected Map<QualifierKind, QualifierKind> initializePolymorphicQualifiers() {
+        Map<QualifierKind, QualifierKind> topToPoly = new TreeMap<>();
+        for (QualifierKind qualifierKind : nameToQualifierKind.values()) {
+            Class<? extends Annotation> clazz = qualifierKind.getAnnotationClass();
+            PolymorphicQualifier polyMetaAnno = clazz.getAnnotation(PolymorphicQualifier.class);
+            if (polyMetaAnno != null) {
+                qualifierKind.isPoly = true;
+                String topName = polyMetaAnno.value().getCanonicalName();
+                if (nameToQualifierKind.containsKey(topName)) {
+                    qualifierKind.top = nameToQualifierKind.get(topName);
+                } else if (topName.equals(PolymorphicQualifier.class.getCanonicalName())) {
+                    if (tops.size() == 1) {
+                        qualifierKind.top = tops.iterator().next();
+                    } else {
+                        throw new UserError(
+                                "Polymorphic qualifier %s did not specify a top annotation class. Tops: [%s]",
+                                qualifierKind, SystemUtil.join(", ", tops));
+                    }
+                } else {
+                    throw new UserError(
+                            "polymorphic qualifier %s's top $s is not a supported qualifier. Found: %s.",
+                            qualifierKind, topName);
+                }
+                topToPoly.put(qualifierKind.top, qualifierKind);
+            }
+        }
+        return topToPoly;
+    }
+
+    /**
+     * For each qualifier in {@code directSuperMap}, initializes {@link QualifierKind#superTypes},
+     * {@link QualifierKind#top} and {@link QualifierKind#bottom}. (Requires tops, bottoms, and
+     * polymorphicQualifiers to be initialized.)
+     *
+     * @param directSuperMap a mapping from a {@link QualifierKind} to a set of its direct super
+     *     qualifiers; created by {@link #createDirectSuperMap()}
+     * @throws UserError if a qualifier isn't a subtype of one of the top qualifiers or if multiple
+     *     tops or bottoms are found for the same hierarchy.
+     */
+    protected void initializeQualifierKindFields(
+            Map<QualifierKind, Set<QualifierKind>> directSuperMap) {
+        for (QualifierKind qualifierKind : nameToQualifierKind.values()) {
+            if (qualifierKind.isPoly) {
+                qualifierKind.superTypes = new TreeSet<>();
+                qualifierKind.superTypes.add(qualifierKind.top);
+            } else {
+                qualifierKind.superTypes = findAllTheSupers(qualifierKind, directSuperMap);
+            }
+        }
+        for (QualifierKind qualifierKind : nameToQualifierKind.values()) {
+            for (QualifierKind top : tops) {
+                if (qualifierKind.isSubtype(top)) {
+                    if (qualifierKind.top == null) {
+                        qualifierKind.top = top;
```

Comment:
As I noted elsewhere, I found it surprising that the `top` field is set for tops elsewhere, and here for all non-tops.  I would put them all in the same place.  Or, document the design.

Reply:
They are all set here now.